network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/pixels/WireguardHandshakeMonitor.kt

```diff
@@ -0,0 +1,165 @@
+import com.duckduckgo.networkprotection.impl.WgProtocol
+import com.squareup.anvil.annotations.ContributesMultibinding
+import java.util.concurrent.atomic.AtomicBoolean
+import javax.inject.Inject
+import kotlin.time.Duration.Companion.minutes
+import kotlin.time.Duration.Companion.seconds
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import logcat.LogPriority.WARN
+import logcat.logcat
+import org.threeten.bp.Instant
+
+@ContributesMultibinding(VpnScope::class)
+class WireguardHandshakeMonitor @Inject constructor(
+    private val wgProtocol: WgProtocol,
+    private val pixels: NetworkProtectionPixels,
+    private val dispatcherProvider: DispatcherProvider,
+    private val networkProtectionState: NetworkProtectionState,
+    private val currentNetworkState: CurrentNetworkState,
+) : VpnServiceCallbacks {
+
+    private val job = ConflatedJob()
+    private val failureReported = AtomicBoolean(false)
+
+    override fun onVpnStarted(coroutineScope: CoroutineScope) {
+        job += coroutineScope.launch {
+            failureReported.set(false)
+            startHandShakeMonitoring()
+        }
+    }
+
+    override fun onVpnReconfigured(coroutineScope: CoroutineScope) {
+        job += coroutineScope.launch {
+            startHandShakeMonitoring()
+        }
+    }
+
+    override fun onVpnStopped(
+        coroutineScope: CoroutineScope,
+        vpnStopReason: VpnStopReason,
+    ) {
+        currentNetworkState.stop()
+        job.cancel()
+    }
+
+    private suspend fun startHandShakeMonitoring() = withContext(dispatcherProvider.io()) {
+        if (networkProtectionState.isEnabled()) {
+            currentNetworkState.start()
+
+            while (isActive && networkProtectionState.isEnabled()) {
+                val nowSeconds = Instant.now().epochSecond
+                val lastHandshakeEpocSeconds = wgProtocol.getStatistics().lastHandshakeEpochSeconds
+                if (lastHandshakeEpocSeconds > 0) {
+                    val diff = nowSeconds - lastHandshakeEpocSeconds
+                    if (diff.seconds.inWholeMinutes > REPORT_TUNNEL_FAILURE_IN_THRESHOLD_MINUTES && currentNetworkState.isConnected()) {
+                        logcat(WARN) { "Last handshake was more than 5 minutes ago" }
+                        // skip if previously reported
+                        if (!failureReported.getAndSet(true)) {
+                            pixels.reportTunnelFailure()
+                        } else {
+                            logcat { "Last handshake was already reported, skipping" }
+                        }
+                    } else {
+                        if (failureReported.getAndSet(false)) {
+                            logcat(WARN) { "Recovered from tunnel failure" }
+                            pixels.reportTunnelFailureRecovered()
```

Comment:
If after `reportTunnelFailure` the device loses connectivity, we are going to report `reportTunnelFailureRecovered`. I don’t think that is a correct signal. Not sure how edge case-y that is though.

Reply:
Good catch, I guess the `else` condition is rather `else if (diff.seconds.inWholeMinutes <= 2)`

Reply:
I am not sure where `2` came from. But yes i think we should check that the diff still to not include the scenario above.